// @ant-design/pro-layout/es/SiderMenu/index.js
import "antd/lib/drawer/style";
import _Drawer from "antd/lib/drawer";
import React from 'react';
import SiderMenu from './SiderMenu';
import { getFlatMenuKeys } from './SiderMenuUtils';

var SiderMenuWrapper = function SiderMenuWrapper(props) {
  var isMobile = props.isMobile,
      menuData = props.menuData,
      collapsed = props.collapsed,
      onCollapse = props.onCollapse;
  var flatMenuKeys = getFlatMenuKeys(menuData);
  return isMobile ? React.createElement(_Drawer, {
    visible: !collapsed,
    placement: "left",
    className: "ant-pro-sider-menu",
    onClose: function onClose() {
      return onCollapse && onCollapse(true);
    },
    style: {
      padding: 0,
      height: '100vh'
    }
  }, React.createElement(SiderMenu, Object.assign({}, props, {
    flatMenuKeys: flatMenuKeys,
    collapsed: isMobile ? false : collapsed
  }))) : React.createElement(SiderMenu, Object.assign({
    className: "ant-pro-sider-menu"
  }, props, {
    flatMenuKeys: flatMenuKeys
  }));
};

SiderMenuWrapper.defaultProps = {
  onCollapse: function onCollapse() {
    return undefined;
  }
};
export default SiderMenuWrapper;
