// @ant-design/pro-layout/es/Footer.js
import "antd/lib/icon/style";
import _Icon from "antd/lib/icon";
import "antd/lib/layout/style";
import _Layout from "antd/lib/layout";
import React, { Fragment } from 'react';
import GlobalFooter from './GlobalFooter';
var Footer = _Layout.Footer;

var FooterView = function FooterView() {
  return React.createElement(Footer, {
    style: {
      padding: 0
    }
  }, React.createElement(GlobalFooter, {
    links: [{
      key: 'Ant Design Pro',
      title: 'Ant Design Pro',
      href: 'https://pro.ant.design',
      blankTarget: true
    }, {
      key: 'github',
      title: React.createElement(_Icon, {
        type: "github"
      }),
      href: 'https://github.com/ant-design/ant-design-pro',
      blankTarget: true
    }, {
      key: 'Ant Design',
      title: 'Ant Design',
      href: 'https://ant.design',
      blankTarget: true
    }],
    copyright: React.createElement(Fragment, null, "Copyright ", React.createElement(_Icon, {
      type: "copyright"
    }), " 2019 \u8682\u8681\u91D1\u670D\u4F53\u9A8C\u6280\u672F\u90E8\u51FA\u54C1")
  }));
};

export default FooterView;
